sql: validate implicit partitioned columns on CREATE UNIQUE INDEX

otan opened the pull request:
This commit modifies CREATE UNIQUE INDEX to check whether the
non-implicit columns in an implicitly partitioned index are unique.

Resolves #59726 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60298)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Is it worth also adding a test to the `regional_by_row` test file?

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60298#-:-MTDfCYB3WlZ77l60mX2:b70hcsy)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/check.go, line 323 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60298#-MTDf-8zD5O1U8_IJvMy:-MTDf-9-ED-BCmucBTSa:b-tr3j11) ([raw file](https://github.com/cockroachdb/cockroach/blob/0640c8b8d234e2b6dd42fed8d77dd37c8ae48fa6/pkg/sql/check.go#L323)):*
> ```Go
> // LIMIT 1  -- if limitResults is set
> //
> func duplicateRowQuery(
> ```

cc @mgartner this is the function we'll need to update to take a predicate


<!-- Sent from Reviewable.io -->

otan commented:
> Is it worth also adding a test to the regional_by_row test file?

unfortunately it's impossible to fail this case with a RBR file, because we're adding the implicit crdb_region in front to an already guaranteed-to-be-unique PK!

TFTR!

bors r=rytaft

rytaft commented:
> unfortunately it's impossible to fail this case with a RBR file, because we're adding the implicit crdb_region in front to an already guaranteed-to-be-unique PK!

Do we not support adding new unique indexes to existing RBR tables yet?





<!-- Sent from Reviewable.io -->

otan commented:
oh i see! that case makes sense

bors r-

otan committed:
sql: validate implicit partitioned columns on CREATE UNIQUE INDEX

This commit modifies CREATE UNIQUE INDEX to check whether the
non-implicit columns in an implicitly partitioned index are unique.

Release note: None

otan commented:
sorry was thinking you meant a separate test case!

in there now

bors r=rytaft

rytaft reviewed (commented):
Thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60298#-:-MTDqzX_8Q8kHJBFRTX7:b6duv8n)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->
